src/llvm-julia-licm.cpp

```diff
@@ -351,6 +366,38 @@ struct JuliaLICM : public JuliaPassContext {
                         MSSAU.insertDef(cast<MemoryDef>(clear_mdef), true);
                     }
                     changed = true;
+                } else if (callee == typeof_func) {
+                    assert(call->arg_size() == 1);
+                    if (!L->isLoopInvariant(call->getArgOperand(0))) {
+                        LLVM_DEBUG(dbgs() << "Failed to hoist typeof because object is not loop invariant: " << *call->getArgOperand(0) << "\n");
+                        continue;
+                    }
+                    ++HoistedTypeof;
+                    moveInstructionBefore(*call, *preheader->getTerminator(), MSSAU, SE);
+                    changed = true;
+                    LLVM_DEBUG(dbgs() << "Hoisted typeof " << *call << "\n");
+                    REMARK([&](){
+                        return OptimizationRemark(DEBUG_TYPE, "Hoist", call)
+                            << "hoisting typeof " << ore::NV("typeof", call);
+                    });
+                } else if (callee == pointer_from_objref_func) {
+                    assert(call->arg_size() == 1);
+                    if (!L->isLoopInvariant(call->getArgOperand(0))) {
+                        LLVM_DEBUG(dbgs() << "Failed to hoist pointer_from_objref because object is not loop invariant: " << *call->getArgOperand(0) << "\n");
+                        continue;
+                    }
+                    if (!preserved_values.contains(call->getArgOperand(0))) {
+                        LLVM_DEBUG(dbgs() << "Failed to hoist pointer_from_objref because object could not be proven to be preserved through the lifetime of the loop: " << *call->getArgOperand(0) << "\n");
+                        continue;
+                    }
```

Comment:
```suggestion
```

That does not seem relevant. `pointer_from_objref` is not a use

Reply:
(pointer_from_objref has identical semantics to ptrtoint, except for being accepted by the gc-verifier pass)

Reply:
Couldn't hoisting it out of a `GC.@preserve` region cause undefined semantics where there previously were defined semantics, if late-gc-lowering thinks that the object doesn't need to be preserved when `pointer_from_objref` is called?

Reply:
It does not need to be preserved when `pointer_from_objref` is called. That function makes no promises about the usefulness of the pointer now (in particular, there is quite lot you are forbidden from assuming about it or using it for), only that it had some value at some past point in time.